docs: add notes (#3260)
statechannels/statechannels: packages/server-wallet/docs/index.md

# What is a state channel wallet?

A node in a state channel network. See [nitro protocol](https://docs.statechannels.org/`).

More concretely, a state channel wallet _is_:

- a queryable store of keys, states and signatures
- a run-loop, triggered by

  - blockchain events
  - API calls
  - messages from counterparties

  that

  - signs states and transactions
  - stores them
  - passes them to external actors

- an event emitter

We usually talk of “The App” as the actor that makes the API calls and binds to the events. The wallet can be a library or run in a separate process to The App (e.g. on a server, or run in an iframe).

A state channel wallet _is not_:

- equipped with a graphical user interface (although this might be a desired plugin)

A state channel wallet has the following _runtime dependencies_:

> - a message delivery service (deferred to The App)

Reviewer: I think this wording is compatible with the status quo API and the "core CM" proposed API.

```suggestion
- a message delivery service (provided by The App)
```

- a blockchain provider (e.g. infura)
- a database (could be a Postgres server or indexedDB)

> The following diagram describes the various components in the architecture. The scope of this document is limited to the components left of the dividing line.

Reviewer: How is this diagram generated? Without an easy way of generating it, it's not that maintainable. (Though, we shouldn't be updating this doc all that often.)

Author: I added a comment about where the diagram was created in this commit https://github.com/statechannels/statechannels/pull/3260/commits/dc92169b273699918abfc673906a0bf3b6071051 . Unfortunately the link won't provide a super easy way to generate this diagram. But it does provide an easy way to generate diagrams of this kind.

Reviewer: That approach does not seem very maintainable. As stated, that's fine when we don't change the diagram.

An alternative like draw.io would allow us to host the "source" of the image in `.git`. Github would render the diagram when displaying a `.md` file. (I don't know what other `.md` renderers will do that, so the ascii diagram definitely has a leg up on portability.)

By [including an `edit-diagram.html`](https://desk.draw.io/support/solutions/articles/16000042371-embed-a-diagram-in-github-markdown) file in our repo, we can even directly edit the diagram, using the `draw.io` UI.

```
                                                    |
                     +-----------------+            |     +--------------+
                     |                 |            |     |              |

>                      |  CHAIN-SERVICE  <------------------>  BLOCKCHAIN  |

Author: Sadly I don't think this will work because the whole diagram is inside a code fence. 

If only github [supported mermaid diagrams](https://github.community/t/feature-request-support-mermaid-markdown-graph-diagrams-in-md-files/1922/105), which could be a nice solution.

         |         |    |          ----------------------->          |
         +---------+    |          |                |     |          |
                        |          |                |     |          |
                        |          |  pushMessage   |     |          |
                        |          <---------------------------+     |
                        |          |                |     |    |     |
                        |          |                |     |    |     |
                        +----------+                |     +----|-----+
                                                    |          |
                                                    | +--------|---------+
                                                    | |  COUNTERPARTY    |
                                                    | |  WALLETS         |
                                                      +------------------+
```

# The responsibilities of a state channel wallet

As in the [Hippocratic oath](https://en.wikipedia.org/wiki/Hippocratic_Oath): _primum non nocere,_ (first, do no harm).

1. Do not lose secrets (private keys)
2. Do not leak secrets (private keys)

> 3. Store relevant states and signatures (Don’t lose any unless they are screened off by kept ones, allow states to be queried)

Reviewer: I don't know what "screened off" means. Is this wording more clear?

```suggestion
3. Store relevant states and signatures (Don’t lose any unless they are superceded by kept ones, allow states to be queried)
```

4. Do not sign states or blockchain transactions unless the App grants permission

Next, do some active things to protect The App’s interests.

5. With implicit permission, allow an exit on chain without losing funds

> 6. Detect and store blockchain challenges (for querying)
> 7. Detect and store changes in channel funding (for querying)

Reviewer: What's the purpose of the (for querying) addition here? I feel like I'd understand 6-7 without it, but adding it makes me confused.
```suggestion
6. Detect and store blockchain challenges
7. Detect and store changes in channel funding
```

> 6. Detect and store blockchain challenges (for querying)
> 7. Detect and store changes in channel funding (for querying)

Reviewer: It might even be more clear not to tell the audience that we store these things. I think this is the _intent_ of the wallet, from an "external point of view".

```suggestion
6. React to blockchain challenges
7. React to changes in channel funding
```

Author: In my mind, reacting to these events is a separate responsibility of the wallet that builds on the more basic responsibility of storing those events. If a partially-functional wallet only stores the challenges but does not otherwise react to them, or if the code that reacts to challenge is somehow buggy, at least users can attempt to recover their funds manually by falling back on the more basic responsibility.

Reviewer: Could one equivalently argue that a wallet which stores states but does not sign new states (for an objective) exhibits the "basic responsibility" of being a wallet? In that case, a user can attempt to fund their app channel by signing ledger updates.

It certainly feels as though the wallet has an _obligation_ to react to a `Challenge` event in some fashion. It may react by calling `checkpoint`, if it detects that the challenge was malicious. It may also react by notifying The App:  "⚠️ !!! A challenge was detected! ⚠️".

Author: Yes I agree, a fully featured wallet should react to challenges. I suppose the direction I am trying to go in, with this ordered list of responsibilities, is that they are built up in a certain order based on "security priority". In that case, reacting to a challenge would be a separately-enumerated responsibility with a lower priority. A wallet should still meet that responsibility in order to be considered feature-complete; but it should not meet it without first meeting all other responsibilities with a higher priority. 

By placing emphasis on _storage_ high up the priority list, I suppose I am trying to discourage wallet builders from implementing features that do not survive a crash/restart cycle.  It's a "walk before you run" approach.

Next, do some active things to allow The App to express itself

8. With explicit permission, sign updates in running application channels
9. With implicit permission, fund channels on chain

Next, allow for advanced funding relationships

> 10. With implicit permission, sign “safe” updates in ledger channels (meaning no funds are lost)

Reviewer: A wallet should also sign updates in eg. guarantor & joint channels.

```suggestion
10. With implicit permission, sign “safe” updates in "auxilliary" funding channels, such as a ledger channel

 "Safe" here means that no funds are lost
```

Author: I think this is an example of where I interpret the terminology slightly differently to others. I tend to think of anything that is not an application channel as a ledger channel.

From the nitro paper: 
> A ledger channel is a channel whose sole purpose is to provide funding to
other channels. 

On the other hand, I think many other people do not interpret it this way. Also your recommendation works under both interpretations. So I went with it!

Reviewer: Is a guarantor channel a ledger channel?

[More detail on responsibilities.](./responsibilities.md)

# Objective-driven Wallet architecture

> In order to fulfil its responsibilities, a state channel wallet must be carefully architected. This section describes one approach that we are pursuing. Let's consider the most important responsibility of a state channels wallet, besides secure storage of secrets and states:

Reviewer: ```suggestion
In order to fulfill its responsibilities, a state channel wallet must be carefully designed. This section describes our approach. Let's consider the most important responsibility of a state channels wallet, besides secure storage of secrets and states:
```

> Signing a state or a blockchain transaction is only ever done with permission.

Here, “explicit permission” shall be implied by an API call. “Implicit permission” is governed by another mechanism that we call _Objectives._ Objectives are a device that allow for

- greater abstraction of the details of nitro protocol
- greater performance and robustness of a responsible wallet
- easier coordination between counterparties
- short-lived request/response cycle over the API

> An objective is a proposal for a goal which The App asserts that it wants to achieve.

> It has a _name_ and _parameters_. The parameters include a _scope_, which is a list of channelIds.

Reviewer: ```suggestion
It has a _name_ and _parameters_. The parameters determine a predefined set of actions that the wallet will take on The App's behalf.
```

> To approve an objective is to grant permission, enabling the wallet to use the private keys it stores as it sees fit, for the channels in its scope for the lifetime of the objective.

> It is _not_ to give permission to spend money, only to sign states and transactions that are overall value-neutral (up to gas fees).

## Implementation: the lifecycle of an objective

### Creation

> Objectives are spawned:
> 
> 1. During request handling, i.e. during an API call. The API call resolves when the objective is committed to the wallet's store.

Reviewer: ```suggestion
Objectives are created:

1. During an API call. The API call resolves when the objective is committed to the wallet's store.
```

> Some objecties are "shared" and involve other participants. They are therefore communicated across "the wire" to other wallets. Therefore we also have objectives being spawned in a wallet:

Reviewer: ```suggestion
Some objecties are "shared" and involve other participants. They are therefore communicated across "the wire" to other wallets. Therefore, objectives can be created by the wallet itself when handling a peer message.
```

## Unrelated comment

Note that this might raise a red flag to a security-oriented engineer: this gives me the ability to DDOS your wallet and jam-pack your store with objectives.

A security engineer might suggest that The App must somehow explicitly grant permission to "create the objective", since ["allowlisting"](https://www.zdnet.com/article/uk-ncsc-to-stop-using-whitelist-and-blacklist-due-to-racial-stereotyping/) is more secure than "blocklisting"*.

\* according to some [random blog](https://blog.finjan.com/blacklisting-vs-whitelisting-understanding-the-security-benefits-of-each/)
> In fact, NIST recommends the use of whitelisting in high-risk security environments, where the integrity of individual or connected systems is critical and takes precedence over any restrictions that users might suffer in their choice or access to software.
but you can find lots of similar opinions.

(Note that there is no reference to NIST's recommendation, but they do have a [Guide to Application Whitelisting](https://csrc.nist.rip/library/alt-SP800-167.pdf).)

> Q: does the app have to first reject the stalled objective? It's scope could be different to the new objective's, so it seems like the only safe thing to do is to release the _entire_ scope of the stalled objective.

### Termination

Objectives can terminate during the run loop (see below), and The App is informed via an event or the return value of an API call triggering that loop.

# The Run loop

The _run loop_ of the wallet looks like this:

1.  Handle an API call, or hear a blockchain event: get all state updates in the store.
2.  UpdateFunding (for any channels that have changed)
3.  CrankToCompletion (for any objectives whose scoped channels have been ‘touched’).
4.  Repeat 2-3 until no more changes
5.  ProcessLedgerQueue (see below)
6.  Repeat 2-5 until no more changes.
7.  Return consolidated updates and states if an API call.

# Updating Funding

TODO

> 

Author: As I said in the PR description, it is my preference that this document is maintained as a record of what we want the wallet to be, rather than what it is.
